xpti/include/xpti_data_types.h

```diff
@@ -0,0 +1,484 @@
+//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//
+#pragma once
+#include <stdint.h>
+#include <string>
+#include <atomic>
+#include <unordered_map>
+
+namespace xpti {
+constexpr int     invalid_id = -1;
+constexpr uint8_t default_vendor = 0;
+
+/// @brief Flag values used by the payload_t structure to mark the information present
+/// @details WHen a payload is created, it is conceivable that only partial information
+/// may be present and these flags are used to indicate the available information. The
+/// hash generator will generate a hash based on the flags set.
+///
+enum class payload_flag_t {
+    NameAvailable        = 1,       ///< The name for the tracepoint is available
+    SourceFileAvailable  = 2,       ///< Source file information available
+    CodePointerAvailable = 4,       ///< Code pointer VA is available
+    LineInfoAvailable    = 8,       ///< Line information available in the payload
+    ColumnInfoAvailable  = 16,      ///< Cilumn information available in payload
+    HashAvailable        = 2 << 16  ///< A hash is already available for this payload
+};
+
+//
+//  Helper macros for creating new tracepoint and
+//  event types
+//
+typedef uint16_t                                     trace_point_t;
+typedef uint16_t                                     event_type_t;
+typedef int32_t                                      string_id_t;
+                                                    
+typedef std::atomic<bool>                            safe_flag_t;
+typedef std::atomic<uint64_t>                        safe_uint64_t;
+typedef std::atomic<uint32_t>                        safe_uint32_t;
+typedef std::atomic<uint16_t>                        safe_uint16_t;
+typedef std::atomic<int64_t>                         safe_int64_t;
+typedef std::atomic<int32_t>                         safe_int32_t;
+typedef std::atomic<int16_t>                         safe_int16_t;
+typedef std::unordered_map<string_id_t, string_id_t> metadata_t;
+
+#define XPTI_EVENT(val)             xpti::event_type_t(val)
```

Comment:
Replace the macros with constexpr functions instead

Reply:
Leaving this one as is for now as I expect this to change a bit before it settles down.